dl/src/main/scala/com/intel/analytics/bigdl/optim/Adadelta.scala

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.reflect.ClassTag
+
+class Adadelta[@specialized(Float, Double) T: ClassTag](implicit ev: TensorNumeric[T])
+  extends OptimMethod[T] {
+
+  /**
+    * Adadelta implementation for SGD: http://arxiv.org/abs/1212.5701
+    *
+    * @param feval     a function that takes a single input (X), the point of a evaluation, and
+    *                  returns f(X) and df/dX
+    * @param parameter the initial point
+    * @param config    a table with hyper-parameters for the optimizer
+    *                  config("learningRate") : learning rate
+    *                  config("learningRateDecay") : learning rate decay
+    *                  config("decayRate"): decayRate, also called interpolation parameter rho
+    *                  config("Epsilon"): for numerical stability
+    * @param state     a table describing the state of the optimizer; after each call the state
+    *                  is modified
+    *                  state("paramVariance") : vector of temporal variances of parameters
+    *                  state("accDelta"): vector of accumulated delta of gradients
+    * @return the new x vector and the function list {fx}, evaluated before the update
+    */
+  override def optimize(feval: (Tensor[T]) => (T, Tensor[T]),
+               parameter: Tensor[T], config: Table, state: Table): (Tensor[T], Array[T]) = {
+
+    val _config = if (config == null) T() else config
+    val _state = if (state == null) _config else state
+
+    val nevals = _state.get[Int]("evalCounter").getOrElse(0)
+    val dr = _config.get[Double]("decayRate").getOrElse(0.9)  // decayRate, alse called rho
+    val eps = _config.get[Double]("Epsilon").getOrElse(1e-10)
+
+    val (fx, dfdx) = feval(parameter)
+
+    val (_paramVariance, _paramStd, _delta, _accDelta) =
+      if (_state.get[Tensor[T]]("paramVariance").isDefined) {
+        (_state.get[Tensor[T]]("paramVariance").get, _state.get[Tensor[T]]("paramStd").get,
+          _state.get[Tensor[T]]("delta").get, _state.get[Tensor[T]]("accDelta").get)
+      } else {
+        (Tensor[T]().resizeAs(dfdx).zero(), Tensor[T]().resizeAs(dfdx).zero(),
+          Tensor[T]().resizeAs(dfdx).zero(), Tensor[T]().resizeAs(dfdx).zero())
+      }
+
+    _paramVariance.mul(ev.fromType[Double](dr)).addcmul(ev.fromType[Double](1-dr), dfdx, dfdx)
+    _paramStd.resizeAs(_paramVariance).copy(_paramVariance).add(ev.fromType[Double](eps)).sqrt()
```

Comment:
why we need the resizeAs here? _paramVariance and _paramStd should be same size with dfdx, right?